thali/thalireplicationmanager.js

```diff
@@ -188,15 +202,62 @@ function syncPeer(peer) {
 
 /* Mux Layer */
 
+function restartMuxServerBridge() {
+  this.stop();
+
+  this.on('stopped', function () {
+    this.start(this._deviceName, this._port, this._dbName);
+  }.bind(this));
+}
+
 function muxServerBridge(tcpEndpointServerPort) {
   var serverPlex = multiplex({}, function(stream, id) {
     var clientSocket = net.createConnection({port: tcpEndpointServerPort});
     stream.pipe(clientSocket).pipe(stream);
   });
 
-  return net.createServer(function(incomingClientSocket) {
+  var server = net.createServer(function(incomingClientSocket) {
+
+    incomingClientSocket.on('error', function (err) {
+      console.log('incoming client socket error %s', err);
+      try {
+        serverPlex.destroy();
+        server.close();
+      } catch (e) {
+        console.log('failed to clean up server and serverPlex');
+      }
+
+      restartMuxServerBridge.call(this);
```

Comment:
These multiple calls to restartMuxServerBridge here and in two more places below make me really nervous. It is at least theoretically possible for say both the incomingClientsocket and the server to error out at the same time. In that case we could end up calling restartMuxServerBridge multiple times. This would seem bad as we would try to stop and start things multiple times. Shouldn't we be protecting against multiple calls to restartMuxServerBridge? Or, more to the point, shouldn't subsequent calls when we already have one restart in progress be treated as thunks?